bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java

```diff
@@ -0,0 +1,296 @@
+    private @NonNullByDefault({}) InputStream inputStream;
+
+    public SDS011Communicator(SDS011Handler thingHandler, SerialPortIdentifier portId) {
+        this.thingHandler = thingHandler;
+        this.portId = portId;
+    }
+
+    /**
+     * Initialize the communication with the device, i.e. open the serial port etc.
+     *
+     * @param mode the {@link WorkMode} if we want to use polling or reporting
+     * @param interval the time between polling or reportings
+     * @return {@code true} if we can communicate with the device
+     * @throws PortInUseException
+     * @throws TooManyListenersException
+     * @throws IOException
+     * @throws UnsupportedCommOperationException
+     */
+    public boolean initialize(WorkMode mode, Duration interval)
+            throws PortInUseException, TooManyListenersException, IOException, UnsupportedCommOperationException {
+        boolean initSuccessful = true;
+
+        SerialPort localSerialPort = portId.open(thingHandler.getThing().getUID().toString(), 2000);
+        localSerialPort.setSerialPortParams(9600, 8, 1, 0);
+
+        outputStream = localSerialPort.getOutputStream();
+        inputStream = localSerialPort.getInputStream();
+
+        if (inputStream == null || outputStream == null) {
+            throw new IOException("Could not create input or outputstream for the port");
+        }
+
+        // wake up the device
+        initSuccessful &= sendSleep(false);
+        initSuccessful &= getFirmware();
+
+        if (mode == WorkMode.POLLING) {
+            initSuccessful &= setMode(WorkMode.POLLING);
+            initSuccessful &= setWorkingPeriod((byte) 0);
+        } else {
+            // reporting
+            initSuccessful &= setWorkingPeriod((byte) interval.toMinutes());
+            initSuccessful &= setMode(WorkMode.REPORTING);
+        }
+
+        // enable listeners only after we have configured the sensor above because for configuring we send and read data
+        // sequentially
+        localSerialPort.notifyOnDataAvailable(true);
+        localSerialPort.addEventListener(this);
+        this.serialPort = localSerialPort;
+
+        return initSuccessful;
+    }
+
+    private @Nullable SensorReply sendCommand(CommandMessage message) throws IOException {
+        byte[] commandData = message.getBytes();
+        if (logger.isDebugEnabled()) {
+            logger.debug("Will send command: {} ({})", Helper.toHexString(commandData), Arrays.toString(commandData));
+        }
+        outputStream.write(commandData, 0, commandData.length);
+        outputStream.flush();
+        try {
+            // Give the sensor some time to handle the command
+            Thread.sleep(500);
+        } catch (InterruptedException e) {
+            logger.warn("Problem while waiting for reading a reply to our command.");
+            Thread.currentThread().interrupt();
+        }
+        SensorReply reply = readReply();
```

Comment:
Wouldn't this already be called implicitly through `serialEvent`? Wouldn't the data on the input stream be gone at that point? Could you explain how this works? I might be misunderstanding how the serial protocol works.

Reply:
I am configuring the sensor with the commands BEFORE I register the listener.

I found it a lot easier to do the protocol stuff in a synchronous manner so I have commands and their replies in one place to check if they belong together and there are no errors.

So this is all done before the calls:
```
localSerialPort.notifyOnDataAvailable(true);
localSerialPort.addEventListener(this);
```

Reply:
Ah! Didn't notice that, makes a lot more sense now, thanks!